Big changes in sftp_handle.py and minor in utils.py and configuration.json

Shettland opened the pull request:
[sftp_handle.py](https://github.com/BU-ISCIII/relecov-tools/blob/develop/relecov_tools/sftp_handle.py) was not verifying the md5 values of the remote files correctly so it had to be fixed. Moreover, I included several testings to extract data from three possible md5checksum configurations: The one from linux `md5sum * > md5checksum.csv` another one from windows which includes headers ("Hash" and "Path") and windows paths (meaning "\" instead of "/") and a last one that has asterisks before the paths coming from third-party softwares.

I also included handling of uncompressed files in order to reduce the space occupied by these downloaded files, and made the code more robust on a general basis by including better error handling and retries for corrupted files (which were previously skipped somehow) and removed the verification or R1/R2 delimiters as it was redundant and obscure.

In order to do so I had to include a few methods in [utils.py](https://github.com/BU-ISCIII/relecov-tools/blob/develop/relecov_tools/utils.py) and some fields into [configuration.json](https://github.com/BU-ISCIII/relecov-tools/blob/develop/relecov_tools/conf/configuration.json) such as the header and sheet for the metadata template in excel format.

These changes should close #242 and close #246 and close #249

saramonzon commented on relecov_tools/sftp_handle.py:
```diff
@@ -511,82 +588,194 @@ def select_target_folders(self):
+            except FileNotFoundError as failed_validation:
+                log.error(failed_validation)
+                stderr.print(f"[red]{failed_validation}")
                 continue
```
how can this continue if you don't have the valid_filedict and meta_file?

Shettland replied:
Because you could have selected multiple folders, there's supposed to be a file for each folder, so if it doesn't exist it goes to the next folder

saramonzon replied:
oki, you are right, my bad!

saramonzon commented on relecov_tools/sftp_handle.py:
```diff
@@ -140,204 +141,269 @@ def close_connection(self):
-                    )
-                    pass
+        sample_data_path = os.path.join(local_folder, sample_data_file)
+        os.rename(metadata_file, os.path.join(local_folder, new_metadata_file))
```
why rename metadata file?

saramonzon replied:
ok, looks good to me :)

saramonzon commented:
There are a lot of issues fixed! Congrats!

The most concerning though I have is that we urgently need a summary for this process output (for all the modules really). We need to now when this finishes how many samples have been downloaded, which ones which errors and which errors. It's very dificut to automatize the communication with the labs if we can't easyly send all errors and warnings, and keep track of everything.

Shettland commented on relecov_tools/sftp_handle.py:
```diff
@@ -511,82 +588,194 @@ def select_target_folders(self):
-                restransmition_data = self.get_files_from_sftp_folder(
-                    folder, req_retransmition
+            remote_md5sum = self.find_remote_md5sum(folder)
+            if remote_md5sum:
```
Yes, it only works for a single md5sum file, I could add the option to use one md5file for each file but I thought it would make the code even more messy

saramonzon replied:
is this described in the manual? I don't mind only one file, but we need to make this as clear as possible, and indicate this in the docs (when we have some of those)

Shettland commented on relecov_tools/sftp_handle.py:
```diff
@@ -379,125 +443,138 @@ def get_sample_fastq_file_names(self, local_folder, meta_f_path):
+            sample_files_dict (dict): same structure as self.get_sample_fastq_file_names
+            local_meta_file (str): location of downloaded metadata excel file
+        """
+        local_meta_file = self.validate_metadata_file(remote_folder, local_folder)
```
Well yes, the method could be namely different, if you wish I can change it to "get_metadata_file"

saramonzon replied:
yes, i think that's clearer

saramonzon replied:
done

Shettland commented on relecov_tools/sftp_handle.py:
```diff
@@ -511,82 +588,194 @@ def select_target_folders(self):
+                )
+            clean_pathlist = [os.path.join(local_folder, fi) for fi in clean_fetchlist]
+            md5_hashes = [
+                relecov_tools.utils.calculate_md5(file) for file in clean_pathlist
```
Yes, they are always calculated here, but if there was an md5sum provided, then those files that had a different md5 won't reach this process because they are deleted

saramonzon replied:
but is it worth it to re-create md5sum files for everyfile even if they have already a correct one? Is not time/computational consuming? It should be created only if they don't have a md5sum file

Shettland replied:
I'll include some kind of handling then

saramonzon replied:
fixed [05aba1f](https://github.com/BU-ISCIII/relecov-tools/pull/247/commits/05aba1f878f16b8f268c5eac4d55be1644dfccd2)

Shettland commented on relecov_tools/sftp_handle.py:
```diff
@@ -140,204 +141,269 @@ def close_connection(self):
+            return False, False
+        # check md5 checksum for each file
+        for f_name in hash_dict.keys():
+            if f_name not in fetched_files:
```
Then those files wouldn't be downloaded, I can try to include some kind of handling for that, but I don't find any reason to download files without any md5 verification when a supposed md5file for verification is provided

saramonzon replied:
but we agreed that we are not requiring md5sum files, only strongly recommended so, this way we are going to lose samples that are missing from the md5sum file. Don't you think we could silently miss those sequence files?

Shettland replied:
You are right, I'll fix that and push the changes to this PR

saramonzon replied:
done

Shettland commented:
> There are a lot of issues fixed! Congrats!
> 
> The most concerning though I have is that we urgently need a summary for this process output (for all the modules really). We need to now when this finishes how many samples have been downloaded, which ones which errors and which errors. It's very dificut to automatize the communication with the labs if we can't easyly send all errors and warnings, and keep track of everything.

That is something that I already created a post in issues for,  I'll try to implement it as soon as possible

Shettland commented on relecov_tools/sftp_handle.py:
```diff
@@ -155,13 +168,23 @@ def list_remote_folders(self, folder_name, recursive=False):
+                raise
             for attribute in attribute_list:
+                # Messy workaround for corrupted folder
+                if "ion_torrent" in attribute.filename:
```
This is a temporal solution that should be removed once that folder is removed from sftp

Shettland commented:
I now introduced some changes that should resolve the things that were pointed out in the conversations.

Nonetheless, I found some bugs during the implementation of these changes that I fixed, e.g. a better handling of only one missing file in the sftp and a custom error type to raise certain errors related with metadata validation.

There are a few temporal solutions that I will add to issues in order to remove them in later commits

Also I introduced a basic validation process for single-end/paired-end reads without the proper number of files, so this now also closes #249
